Add blog posts for React 19 Beta (#6778)
reactjs/react.dev: src/content/blog/2024/04/25/react-19-upgrade-guide.md

For a list of changes in 18.3 see the [Release Notes](https://github.com/facebook/react/blob/main/CHANGELOG.md).

</Note>

In this post, we will guide you through the steps for upgrading libraries to React 19 beta:

- [Installing](#installing)
- [Breaking Changes](#breaking-changes)
- [New Deprecations](#new-deprecations)
- [Notable Changes](#notable-changes)
- [TypeScript Changes](#typescript-changes)
- [Changelog](#changelog)

If you'd like to help us test React 19, follow the steps in this upgrade guide and [report any issues](https://github.com/facebook/react/issues/new/choose) you encounter. For a list of new features added to React 19 beta, see the [React 19 release post](/blog/2024/04/25/react-19).

---
## Installing {/*installing*/}

<Note>

#### New JSX Transform is now required {/*new-jsx-transform-is-now-required*/}

We introduced a [new JSX transform](https://legacy.reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html) in 2020 to improve bundle size and use JSX without importing React. In React 19, we're adding additional improvements like using ref as a prop and JSX speed improvements that require the new transform.

> If you the new transform is not enabled, you will see this warning:

Reviewer: ```suggestion
If the new transform is not enabled, you will see this warning:
```

</Note>

To install the latest version of React and React DOM:

```bash
npm install react@beta react-dom@beta
```

If you're using TypeScript, you also need to update the types. Once React 19 is released as stable, you can install the types as usual from `@types/react` and `@types/react-dom`.  During the beta period, the types are available in different packages which need to be enforced in your `package.json`:

```json
{
  "dependencies": {
    "@types/react": "npm:types-react@alpha",
    "@types/react-dom": "npm:types-react-dom@alpha"
  },
  "overrides": {
    "@types/react": "npm:types-react@alpha",
    "@types/react-dom": "npm:types-react-dom@alpha"
  }
}
```

We're also including a codemod for the most common replacements. See [Removed TypeScript Types](#removed-deprecated-typescript-types) below.

## Breaking Changes {/*breaking-changes*/}

### Errors in render are not re-thrown {/*errors-in-render-are-not-re-thrown*/}

> In previous versions of React, errors thrown during render were caught in rethrown. In DEV, we would also log to `console.error`, resulting in duplicate error logs. 

Reviewer: Not sure this suggestion is right....

```suggestion
In previous versions of React, errors thrown during render were caught and rethrown. In DEV, we would also log to `console.error`, resulting in duplicate error logs. 
```

class Parent extends React.Component {
  render() {
    return (
      <FooContext value='bar'>
        <Child />
      </FooContext>
    );
  }
}

class Child extends React.Component {
  static contextType = FooContext;

  render() {
    return <div>{this.context}</div>;
  }
}
```

#### Removed: string refs {/*removed-string-refs*/}
String refs were deprecated in [March, 2018 (v16.3.0)](https://legacy.reactjs.org/blog/2018/03/27/update-on-async-rendering.html).

Class components supported string refs before being replaced by ref callbacks due to [multiple downsides](https://github.com/facebook/react/issues/1373). In React 19, we're removing string refs to make React simpler and easier to understand.

If you're still using string refs in class components, you'll need to migrate to ref callbacks:

```js
// Before
class MyComponent extends React.Component {
  componentDidMount() {
    this.refs.input.focus();
  }

  render() {
    return <input ref='input' />;
  }
}
```

```js
// After
class MyComponent extends React.Component {
  componentDidMount() {
    this.input.focus();
  }

  render() {
    return <input ref={input => this.input = input} />;
  }
}
```

<Note>

> To help with the migration, we've created a [react-codemod](/todo) that will automatically update your code to use ref callbacks.

Reviewer 2: ```suggestion
To help with the migration, we've created a [react-codemod](https://github.com/reactjs/react-codemod/#string-refs) that will automatically update your code to use ref callbacks.
```

once https://github.com/reactjs/react-codemod/pull/309 is stamped and released

> TODO: instructions.

Reviewer 3: 👀
